packages/render/src/renderAsync.ts

```diff
@@ -0,0 +1,66 @@
+import { convert } from "html-to-text";
+import pretty from "pretty";
+import { type ReactNode } from "react";
+import { renderToReadableStream, renderToStaticMarkup } from "react-dom/server";
+import { type ReadableStream } from "node:stream/web";
+
+export default async function renderToString(children: ReactNode) {
+  const stream = await renderToReadableStream(children);
+
+  const html = await readableStreamToString(
+    // ReactDOMServerReadableStream behaves like ReadableStream
+    // in modern edge runtimes but the types are not compatible
+    stream as unknown as ReadableStream
+  );
+
+  return (
+    html
+      // Remove leading doctype becuase we add it manually
+      .replace(/^<!DOCTYPE html>/, "")
+      // Remove empty comments to match the output of renderToStaticMarkup
+      .replace(/<!-- -->/g, "")
+  );
+}
+
+async function readableStreamToString(readableStream: ReadableStream) {
+  const chunks: Uint8Array[] = [];
+
+  for await (const chunk of readableStream) {
+    chunks.push(chunk);
+  }
+
+  return Buffer.concat(chunks).toString("utf-8");
```

Comment:
Buffer is a node api right? It would be nice to not have a dependency on that. I'm planning on running this on Vercel's Edge runtime.

Reply:
When I bring this code into my project, I get an error on that for loop
<img width="778" alt="Screen Shot 2023-03-24 at 11 47 14 PM" src="https://user-images.githubusercontent.com/621286/227701732-8f4e9a64-8788-4e76-bb45-e7fbe6b7c09c.png">

Reply:
I found this example code that works.  I don't like the while loop, but it's working and I've moved onto other things:
```
async function concatStringStream(stream: ReadableStream): Promise<string> {
  const reader = stream.getReader();
  const chunks: Array<any> = [];
  while (true) {
    // The `read()` method returns a promise that
    // resolves when a value has been received.
    const { done, value } = await reader.read();
    // Result objects contain two properties:
    // `done`  - `true` if the stream has already given you all its data.
    // `value` - Some data. Always `undefined` when `done` is `true`.
    if (done) {
      const buffer = Buffer.concat(chunks);
      return buffer.toString('utf-8');
    }
    chunks.push(value);
  }
}
```

Reply:
> Buffer is a node api right? It would be nice to not have a dependency on that. I'm planning on running this on Vercel's Edge runtime.

The code I post is working for me in a Vercel Edge function.  It seems Buffer is ubiquitous.

Reply:
Here's the rest of the fns related to generating both html and plain text emails using that function:
```typescript
const renderAsPlainText = async (html: string) => {
  return convert(html, {
    selectors: [
      { selector: 'img', format: 'skip' },
      { selector: '#__react-email-preview', format: 'skip' }
    ]
  });
};
const render = async (
  component: React.ReactElement,
  options: RenderOptions | undefined | null
) => {
  const markup = await concatStringStream(
    await ReactDOMServer.renderToReadableStream(component)
  );

  if (options?.plainText) {
    return renderAsPlainText(markup);
  }
  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${markup}`;
  if (options?.pretty) {
    return pretty(document);
  }
  return document;
};

export const renderHtmlEmail = async (args: EmailArgs): Promise<string> =>
  render(<Email {...args} />, {
    pretty: true
  });

export const renderTextEmail = async (args: EmailArgs): Promise<string> =>
  render(<Email {...args} />, {
    plainText: true
  });
```

Reply:
> When I bring this code into my project, I get an error on that for loop
> 
> <img width="778" alt="Screen Shot 2023-03-24 at 11 47 14 PM" src="https://user-images.githubusercontent.com/621286/227701732-8f4e9a64-8788-4e76-bb45-e7fbe6b7c09c.png">
> 
> 

Dumb question, but do you have imported it from 'node:stream/web'?

Reply:
> Buffer is a node api right? It would be nice to not have a dependency on that. I'm planning on running this on Vercel's Edge runtime.

Buffer should be available in every modern runtime including edge runtimes afaik.

Reply:
> Buffer should be available in every modern runtime including edge runtimes afaik

I don't think that is right.

https://vercel.com/docs/concepts/functions/edge-functions/edge-runtime
https://nodejs.org/api/buffer.html
https://common-min-api.proposal.wintercg.org/

Reply:
> > When I bring this code into my project, I get an error on that for loop 
> Dumb question, but do you have imported it from `node:stream/web`?

It works if I import that, but I doubt that's available on the edge because it appears to be a Node library.

The type that ReactDOMServer.renderToReadableStream returns is `ReactDOMServer.ReactDOMServerReadableStream` 

<img width="527" alt="Screen Shot 2023-03-28 at 2 08 24 PM" src="https://user-images.githubusercontent.com/621286/228367465-60444eae-bdf9-46e2-b761-4c284e7ea773.png">

and if we type the argument to `readableStreamToString` with that then it gives the same error I showed before:

<img width="861" alt="Screen Shot 2023-03-28 at 2 06 28 PM" src="https://user-images.githubusercontent.com/621286/228366929-6de2f69c-dab4-4b6f-971c-429c868645ee.png">

Reply:
> It works if I import that, but I doubt that's available on the edge because it appears to be a Node library.
> 
> The type that ReactDOMServer.renderToReadableStream returns is ReactDOMServer.ReactDOMServerReadableStream

`ReadableStream` is available in Vercel Edge functions, check the docs.
https://vercel.com/docs/concepts/functions/edge-functions/edge-runtime#stream-apis
https://common-min-api.proposal.wintercg.org/#index

`ReactDOMServer.ReactDOMServerReadableStream` appears to be a standard `ReadableStream`, but the types are not compatible.